### PR TITLE
Implement scatter reductions (CUDA), remove divide/subtract

### LINKED ISSUE #33394: Non-deterministic scatter reduction algorithms for scatter operations for CUDA (sum, subtract, divide, multiply).
## 🚀 Feature

CUDA equivalent of https://github.com/pytorch/pytorch/issues/33389

New issue for the sake of clarity and brevity.

### PR DESCRIPTION
**Pull request by v0dro**
Fixes https://github.com/pytorch/pytorch/issues/33394 .

This PR does two things:
1. Implement CUDA scatter reductions with revamped GPU atomic operations.
2. Remove support for divide and subtract for CPU reduction as was discussed with @ngimel .

I've also updated the docs to reflect the existence of only multiply and add.

### CODE REVIEW

File: aten/src/ATen/native/cuda/AtomicOps.cuh [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+
+    do {
+      assumed = old;
+      old_bytes = is_32_align ? old >> 16 : old & 0xffff;
+      // preserve size in initial cast. Casting directly to uint32_t pads
+      // negative signed values with 1's (e.g. signed -1 = unsigned ~0).
+      newval = func(val, old_bytes);//static_cast<uint16_t>(THCNumerics<T>::add(val, old_bytes));
+      newval = is_32_align ? (old & 0xffff) | (newval << 16) : (old & 0xffff0000) | newval;
+      old = atomicCAS(address_as_ui, assumed, newval);
+    } while (assumed != old);
```
Comment (aocsa) [inline aten/src/ATen/native/cuda/AtomicOps.cuh:63]: Could you please provide a reference for these implementations. If I understand well, these implementation extends `atomicAdd` function for types that are not covered by specific CUDA architecture/version.

**Review by aocsa — COMMENTED** [review]
Nice job! @v0dro, I really enjoyed reading/reviewing this PR about scatter reduce on CUDA. I put a couple of general comments and questions.

File: aten/src/ATen/native/cuda/AtomicOps.cuh (hunk 1 above)
Comment (v0dro) [reply aten/src/ATen/native/cuda/AtomicOps.cuh:63]: They're derived from THCAtomics which was already present in the pytorch source code. In time, I think these routines can replace the ones in THCAtomics since these are more generic. 

https://github.com/pytorch/pytorch/blob/master/aten/src/THC/THCAtomics.cuh

File: aten/src/ATen/native/cuda/AtomicOps.cuh [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+    AtomicIntegerop<int16_t, sizeof(int16_t)>()(address, val,
+                                                [](int16_t val, uint32_t old_byte) {
+                                                  return static_cast<uint16_t>(THCNumerics<int16_t>::add(val, old_byte));
+                                                });
+  }
+
+  inline __device__ int32_t operator() (int32_t * address, int32_t val) {
+    return atomicAdd(address, val);
+  }
+
```
[the remark this answers is not in the document]
Comment (v0dro) [reply aten/src/ATen/native/cuda/AtomicOps.cuh:196]: Yes this is type-size depenendent specialization. I can add some tests for these atomic routines.

**Comment by v0dro** [comment]
Anybody has ideas on this failure? Not even the Docker can reproduce it.

File: aten/src/ATen/native/cuda/AtomicOps.cuh [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/AtomicOps.cuh:1]: So now we have THCAtomics.cuh and this file, which duplicates THCAtomic.cuh functionality/implementation for add. This is pretty confusing. 
Given that we only dispatch floating types, implementing multiple ops for all integer variants seems excessive? What's needed is atomicMul for bfloat16, half, float and double, right, and that can probably go to THCAtomics.

File: aten/src/ATen/native/cuda/AtomicOps.cuh [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+    AtomicIntegerop<int64_t, sizeof(int64_t)>()(address, val,
+                                                [](int64_t val, unsigned long long old) {
+                                                  return val * (int64_t)old;
+                                                });    
+  }
+
+  inline __device__ at::Half operator() (at::Half * address, at::Half val) {
+    return AtomicFPOp<at::Half>()(address, val,
+                                  [](at::Half bsum, at::Half val) {
+                                    return THCNumerics<at::Half>::add(bsum, val);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/AtomicOps.cuh:300]: looks like copy paste, should be mul instead of add. Strange that it is not caught by tests.

File: aten/src/ATen/native/cuda/AtomicOps.cuh [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+
+  inline __device__ at::BFloat16 operator() (at::BFloat16 * address, at::BFloat16 val) {
+    return AtomicFPOp<at::BFloat16>()(address, val,
+                                      [](at::BFloat16 bsum, at::BFloat16 val) {
+                                        return THCNumerics<at::BFloat16>::add(bsum, val);
+                                      });    
+  }
+
+  inline __device__ void operator() (bool * address, bool val) {
+    *address = address && val;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/AtomicOps.cuh:312]: also does not look right, and is not needed as we are not dispatching bool

File: aten/src/ATen/native/cuda/AtomicOps.cuh (hunk 3 above)
Comment (v0dro) [reply aten/src/ATen/native/cuda/AtomicOps.cuh:1]: I thought I'd future proof the implementation by implementing for all the data types. I've just added CUDA-specific tests using C++ for all the data types.

IMO the excessive bloat caused in the THCAtomics file justifies putting this in a separate file and changing all the usages of THCAtomics `gpuAtomicAdd` to use the new implementation in a later PR.

File: aten/src/ATen/native/cuda/AtomicOps.cuh (hunk 5 above)
Comment (v0dro) [reply aten/src/ATen/native/cuda/AtomicOps.cuh:312]: An implementation for bool has to exist in order to make the `cuda_scatter_gather_base_kernel` in the ScatterGatherKernel.cu file generic enough to work with assignment and mul/add.

Comment (ngimel) [reply aten/src/ATen/native/cuda/AtomicOps.cuh:312]: I don't see why - cuda_scatter_gather_base_kernel should not be dispatched for bool.

File: aten/src/ATen/native/cuda/AtomicOps.cuh (hunk 4 above)
Comment (v0dro) [reply aten/src/ATen/native/cuda/AtomicOps.cuh:300]: Fixed.

File: aten/src/ATen/native/cuda/AtomicOps.cuh (hunk 3 above)
Comment (ngimel) [reply aten/src/ATen/native/cuda/AtomicOps.cuh:1]: Not if we limit the extensions to only the necessary functions. 
There are a few problems with switching things to the new implementations:
1) There are extensions that depend on THCAtomics, changing them for no good reason is tedious
2) (fixable) functions here are `void` instead of returning old value, and some extensions rely on returning old values. 
3) there's fastAtomicAdd implementation in kernelUtils.cuh, which is immensely better for halfs, if we want to go through the trouble of switching all atomic uses, we might as well switch to it
4) the integer implementations are very slow, and only cause user confusion (#42109), there's no point having them just for the sake of having them
In short, I think we should be adding only functions and specializations we really need, and we better keep them in the same file, so that it's easier to keep track of what we have and what we don't.

Comment (v0dro) [reply aten/src/ATen/native/cuda/AtomicOps.cuh:1]: Ok makes sense. I'll revert to adding the multiplication to the old file and use the old functions for addition.

File: aten/src/ATen/native/cuda/AtomicOps.cuh (hunk 5 above)
Comment (v0dro) [reply aten/src/ATen/native/cuda/AtomicOps.cuh:311]: It uses the same code for the assignment operator as well as add/mult. Boolean assignment has been around before too.

**Comment by v0dro** [comment]
@ngimel could you please have a look at this? All tests except the linting pass.

**Comment by v0dro** [comment]
AFAIK the ROCM builds fail for no fault of mine.

File: aten/src/ATen/test/cuda_atomic_ops_test.cu [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  T *a, *sum, *answer, *ad, *sumd;
+  
+  a = (T*)malloc(arraysize * sizeof(T));
+  sum = (T*)malloc(arraysize * sizeof(T));
+  answer = (T*)malloc(arraysize * sizeof(T));
+
+  for (int i = 0; i < arraysize; ++i) {
+    a[i] = 2;
+    sum[i] = 2;
+    answer[i] = 8*factor;
```
Comment (ngimel) [inline aten/src/ATen/test/cuda_atomic_ops_test.cu:78]: This is not the correct expression, factor should be exponent, not multiplicand (e.g. if factor was 3, correct answer would be 16, not 8*3)

File: aten/src/THC/THCAtomics.cuh [hunk 7]
```diff
@@ -247,4 +281,57 @@ static inline __device__ void atomicAdd(bool *address, bool val) {
...
+    old = atomicCAS(address_as_ull, assumed,
+                    __float_as_int(val *
+                                   __int_as_float(assumed)));
+
+    // Note: uses integer comparison to avoid hang in case of NaN (since NaN != NaN)
+  } while (assumed != old);
+}
+
+template<typename T>
+inline __device__ void gpuAtomicMul (c10::complex<T> * address, c10::complex<T> val) {
```
Comment (ngimel) [inline aten/src/THC/THCAtomics.cuh:333]: This is not complex multiplication and properly enabling complex atomic multiplication is very challenging. I suggest not enabling scatter_reduce with mul for complex (and same for future work, we don't enable complex min and max otherwise, so we should not enable them for scatter operations). 
As an aside, complex gpuAtomicAdd as it exists now it also not atomic, so should not be called that, but no action required for this PR.

**Review by ngimel — COMMENTED** [review]
Sorry for delay, generally looks good, please disable multiplication for complex inputs, as it's a lot of work with unclear usecase.

File: aten/src/THC/THCAtomics.cuh (hunk 7 above)
Comment (v0dro) [reply aten/src/THC/THCAtomics.cuh:333]: Ah my bad. Should not missed this one sorry. Will change it.

**Comment by v0dro** [comment]
@ngimel something went wrong with the merge with master branch. I will open a new PR and link it here.

**Comment by v0dro** [comment]
I don't know why it requested review from so many people.

**Comment by ngimel** [comment]
So what you are doing is dispatching for all types (including bool and complex) but bool and complex will trigger device assert. Maybe not bool, because you check for floating + complex types earlier, but definitely complex. Please do the following
1) adjust your dispatch macros so that kernels for unsupported types aren't dispatched. That may require code duplication, because supported types are different for addition and multiplication, but it's still safer situation than now. After that is done, you won't need atomic op specialization that would trigger device assert, and that's how it should be. Imagine someone using atomic ops, compiling their code successfully, only to discover device assert at runtime. 
2) make sure that runtime error is thrown if operation is called on an unsupported type (this is not happening currently with multiply)
3) add tests checking that.

**Comment by ngimel** [comment]
@v0dro any update on this?

**Comment by v0dro** [comment]
Working on it today and will send an update in a few hours.

File: test/test_torch.py [hunk 8]
```diff
@@ -12507,71 +12503,55 @@ def test_put_empty(self, device):
...
+                               [12, 2, 2, 2],
+                               [2, 2, 2, 2]], device=device, dtype=dtype), "multiply"),
+            ]
 
-        for input, src, result, operation in test_data:
-            input.scatter_(0, index, src, reduce=operation)
-            self.assertEqual(input, result)
+            for input, src, result, operation in test_data:
+                if operation == "multiply" and torch.is_complex(input):
+                    continue
```
Comment (ngimel) [inline test/test_torch.py:12528]: can you please check do `self.assertRaisesRegexp` here and in similar places?

**Review by ngimel — COMMENTED** [review]
This is almost ready now, please test that scatter_reduce with multiply and complex types throws runtime error. Thank you!

File: test/test_torch.py (hunk 8 above)
Comment (v0dro) [reply test/test_torch.py:12528]: I'm doing this in a separate test here: https://github.com/pytorch/pytorch/pull/41977/files#diff-9996665f82f52030836eb8657057cfadR12591

File: test/test_torch.py [hunk 9]
```diff
@@ -12586,36 +12566,37 @@ def test_scatter_add_non_unique_index(self, device):
...
+                              dtype=dtype).repeat(1, width), "multiply"),
+            ]
 
+            for input, src, result, operation in test_data:
+                if operation == "multiply" and torch.is_complex(input):
+                    continue
+                input.scatter_(0, index, src, reduce=operation)
+                self.assertEqual(input, result, msg=f"result: {result} input: {input} method: {str(operation)}")
+    @onlyCUDA
+    def test_scatter_reduce_multiply_unsupported_dtypes(self, device):
```
[the remark this answers is not in the document]
Comment (v0dro) [reply test/test_torch.py:12592]: Add is supported for all types. I think simply adding the decorators should be enough since we have the same test for add and multiply with just an if condition that skips integer/complex dtype testing for multiply.

**Comment by v0dro** [comment]
@ngimel this fails for XLA and I think the reason is that the decorators don't support that device. Should i make a new decorator or make these tests exclusive for CUDA and CPU?

**Comment by ngimel** [comment]
You should have @dtypes decorator (that will apply to XLA), and @dtypesIfCPU, @dtypesIfCUDA that you currently have. There are examples in the test suite that have all 3. But having just @dtypeIfCPU and @dtypesIfCUDA is an error, you always need plain @dtypes.

**Comment by ngimel** [comment]
xla error is real, can you please skip part of the test checking for runtime error on xla? Also, rocm build interestingly times out on scatter test, so it's hard to say if it's related or not.

**Comment by ngimel** [comment]
Hm, rocm build times out on a scatter test, that's worrying. Can you disable it on rocm?

**Comment by v0dro** [comment]
OK everything passing except the facebook internal tests.

**Review by ngimel — APPROVED** [review]

**Comment by malfet** [comment]
Why changes to submodules were pulled as part of this PR? (namely, it reverted #44706)

**Comment by sbb-gh** [comment]
Is there an option to ignore/sum over several of the output dimensions?
https://discuss.pytorch.org/t/scatter-add-reduce-output-dimensions-shape/100427
